dhis-2/dhis-services/dhis-service-acl/src/main/java/org/hisp/dhis/security/acl/DefaultAclService.java

```diff
@@ -530,99 +618,97 @@ public <T extends IdentifiableObject> List<ErrorReport> verifySharing(T object,
       }
     }
 
-    boolean canMakePublic = canMakePublic(user, object);
-    boolean canMakePrivate = canMakePrivate(user, object);
-    boolean canMakeExternal = canMakeExternal(user, object);
+    boolean canMakePublic = canMakePublic(userDetails, object);
+    boolean canMakePrivate = canMakePrivate(userDetails, object);
+    boolean canMakeExternal = canMakeExternal(userDetails, object);
 
-    if (object.getSharing().isExternal()) {
-      if (!canMakeExternal) {
-        errorReports.add(
-            new ErrorReport(
-                object.getClass(), ErrorCode.E3006, user.getUsername(), object.getClass()));
-      }
+    if (object.getSharing().isExternal() && !canMakeExternal) {
+      errorReports.add(
+          new ErrorReport(object.getClass(), ErrorCode.E3006, userDetails, object.getClass()));
     }
 
-    errorReports.addAll(verifyImplicitSharing(user, object));
+    errorReports.addAll(verifyImplicitSharing(userDetails, object));
 
     if (AccessStringHelper.DEFAULT.equals(object.getSharing().getPublicAccess())) {
       if (canMakePublic || canMakePrivate) {
         return errorReports;
       }
 
       errorReports.add(
-          new ErrorReport(
-              object.getClass(), ErrorCode.E3009, user.getUsername(), object.getClass()));
+          new ErrorReport(object.getClass(), ErrorCode.E3009, userDetails, object.getClass()));
     } else {
       if (canMakePublic) {
         return errorReports;
       }
 
       errorReports.add(
-          new ErrorReport(
-              object.getClass(), ErrorCode.E3008, user.getUsername(), object.getClass()));
+          new ErrorReport(object.getClass(), ErrorCode.E3008, userDetails, object.getClass()));
     }
 
     return errorReports;
   }
 
   private <T extends IdentifiableObject> Collection<? extends ErrorReport> verifyImplicitSharing(
-      User user, T object) {
+      UserDetails userDetails, T object) {
     List<ErrorReport> errorReports = new ArrayList<>();
     Schema schema = schemaService.getSchema(HibernateProxyUtils.getRealClass(object));
 
     if (!schema.isImplicitPrivateAuthority()
-        || checkMetadataSharingPermission(user, object, Permission.WRITE)) {
+        || checkMetadataSharingPermission(userDetails, object, Permission.WRITE)) {
       return errorReports;
     }
 
     if (AccessStringHelper.DEFAULT.equals(object.getSharing().getPublicAccess())) {
       errorReports.add(
-          new ErrorReport(
-              object.getClass(), ErrorCode.E3001, user.getUsername(), object.getClass()));
+          new ErrorReport(object.getClass(), ErrorCode.E3001, userDetails, object.getClass()));
     }
 
     return errorReports;
   }
 
-  private boolean haveOverrideAuthority(User user) {
+  private boolean haveOverrideAuthority(UserDetails user) {
+    return user == null || user.isSuper();
```

Comment:
is this similar to the old pattern we're trying to remove? (treating null user as super, now treating null userDetails as super?)